README.md
# Afterglow

[![project chat](https://img.shields.io/badge/chat-on%20zulip-brightgreen)](https://deep-symmetry.zulipchat.com/#narrow/stream/318697-afterglow)
 <image align="right" width="275" height="255"
 src="doc/modules/ROOT/assets/images/Afterglow-logo-padded-left.png"> <br/><br/> An
 environment supporting
 [live coding](https://en.wikipedia.org/wiki/Live_coding) for the
 creation of algorithmic light shows in [Clojure](http://clojure.org),
 leveraging the
 [Open Lighting Architecture](https://www.openlighting.org/ola/) with
 the help of
 [ola-clojure](https://github.com/Deep-Symmetry/ola-clojure#ola-clojure),
 [wayang](https://github.com/Deep-Symmetry/wayang#wayang),
 [beat-link](https://github.com/Deep-Symmetry/beat-link), and pieces of
 the [Overtone](http://overtone.github.io) toolkit. Beyond building on
 pieces of Overtone, the entire Afterglow project was
 [inspired](https://vimeo.com/22798433) by it.

[![License](https://img.shields.io/badge/License-Eclipse%20Public%20License%202.0-blue.svg)](#license)

### Documentation Overview

This page provides an introduction in how to install and use
Afterglow. The [Developer
Guide](https://afterglow-guide.deepsymmetry.org) goes much deeper,
and there is also [API
documentation](http://afterglow-guide.deepsymmetry.org/api/). For more
interactive help, the [Afterglow stream on
Zulip](https://deep-symmetry.zulipchat.com/#narrow/stream/318697-afterglow) is the place to start,
and if you want to see (or contribute to) more structured and lasting
community-driven documentation, there is also a project
[wiki](https://github.com/Deep-Symmetry/afterglow/wiki).

## Why Explore Afterglow?

> tl;dr&mdash;show me? Check out the
> [Show Control pics](https://afterglow-guide.deepsymmetry.org/afterglow/README.html#show-control)
> and [performance video](https://afterglow-guide.deepsymmetry.org/afterglow/videos.html).

As suggested by the live-coding orientation mentioned above, which is
designed to let you inject your own code right into the frame
rendering process, Afterglow takes a very different approach to
controlling light shows than other software. It won&rsquo;t be right
for everyone, but will be extremely compelling to a particular niche.
The early stages of its [rendering
loop](https://afterglow-guide.deepsymmetry.org/afterglow/rendering_loop.html#the-rendering-loop)
can offer higher levels of abstraction than the usual DMX [channel
value](https://afterglow-guide.deepsymmetry.org/afterglow/effects.html#channel-effects)
or [fixture
function](https://afterglow-guide.deepsymmetry.org/afterglow/effects.html#function-effects)
(although those are fully supported too):

* You can express your desired results in terms of an abstract
  [color](https://afterglow-guide.deepsymmetry.org/afterglow/effects.html#color-effects),
  including support for the hue-saturation-lightness model, which is
  great for algorithmic looks, and have it translated to whatever
  color channels (or color wheel) your fixture supports.

* Groups of moving heads can be told to face particular
  [directions](https://afterglow-guide.deepsymmetry.org/afterglow/effects.html#direction-effects)
  by specifying parameterized vectors, or to
  [aim](https://afterglow-guide.deepsymmetry.org/afterglow/effects.html#aim-effects)
  at a particular point in space, and Afterglow figures out how to
  translate that into DMX control values given its understanding of
  the
  [fixture](https://afterglow-guide.deepsymmetry.org/afterglow/fixture_definitions.html)
  and
  [where](https://afterglow-guide.deepsymmetry.org/afterglow/show_space.html),
  and at what angle, you hung it.

* There are a variety of
  [oscillators](https://afterglow-guide.deepsymmetry.org/afterglow/oscillators.html)
  which can efficiently drive effect parameters.

* You can also create
  [complex effects](https://afterglow-guide.deepsymmetry.org/afterglow/effects.html#complex-effects),
  with
  [adjustable parameters](https://afterglow-guide.deepsymmetry.org/afterglow/parameters.html)
  that can be controlled through a rich binding to an
  [Ableton Push](https://afterglow-guide.deepsymmetry.org/afterglow/push2.html) or
  [Novation Launchpad family](https://afterglow-guide.deepsymmetry.org/afterglow/launchpad.html)
  controller, or via
  [Open Sound Control](https://afterglow-guide.deepsymmetry.org/afterglow/mapping_sync.html#open-sound-control)
  (OSC)&mdash;even wirelessly from a tablet or smartphone.

* The timing of effects is pervasively influenced by a deep notion of
  [musical time](https://afterglow-guide.deepsymmetry.org/afterglow/metronomes.html),
  with support for synchronization via
  [MIDI clock](https://afterglow-guide.deepsymmetry.org/afterglow/mapping_sync.html#syncing-to-midi-clock),
  [Traktor Beat Phase](https://afterglow-guide.deepsymmetry.org/afterglow/mapping_sync.html#syncing-to-traktor-beat-phase),
  or Pioneer
  [Pro DJ Link](https://afterglow-guide.deepsymmetry.org/afterglow/mapping_sync.html#syncing-to-pro-dj-link)
  beat grids.

* You can even host Afterglow within
  [Cycling ‘74’s Max](https://cycling74.com/) visual interactive
  environment.

If any of this sounds interesting to you, read on to see how to get
started!

## Table of Contents

  * [Why Explore Afterglow?](#why-explore-afterglow)
  * [Installation](#installation)
  * [Status](#status)
  * [Usage](#usage)
  * [Troubleshooting](#troubleshooting)
  * [Bugs](#bugs)
  * [What Next?](#what-next)
  * [Release Checklist](#release-checklist)
  * [Tasks](#tasks)
  * [License](#license)

## Installation

1. [Install OLA](https://www.openlighting.org/ola/getting-started/downloads/).
   (On the Mac I recommend using [Homebrew](http://brew.sh) which lets you simply
   `brew install ola`). Once you launch the `olad` server you can
   interact with its embedded
   [web server](http://localhost:9090/ola.html), which is very helpful
   in seeing whether anything is working; you can even watch live DMX
   values changing.

   > :wrench: If you are installing Afterglow on Windows, see the
   > [Wiki discussion](https://github.com/Deep-Symmetry/afterglow/wiki/Questions#ola-and-windows)
   > about OLA options.

2. For now set up a Clojure project using [Leiningen](http://leiningen.org).

3. Add this project as a dependency:
   [![Clojars Project](https://img.shields.io/clojars/v/afterglow.svg)](http://clojars.org/afterglow)

> :wrench: If you were using older releases of Afterglow and installed
> [CoreMIDI4J](https://github.com/DerekCook/CoreMidi4J) in
> `/Library/Java/Extensions`, you need to remove it, because Afterglow
> now embeds an improved version and uses it when necessary.

If you want to run Afterglow as a standalone executable, you can
download the executable &uuml;berjar from the
[releases](https://github.com/Deep-Symmetry/afterglow/releases) page.
[![&uuml;berjar](https://img.shields.io/github/downloads/Deep-Symmetry/afterglow/total.svg)](https://github.com/Deep-Symmetry/afterglow/releases)

For an example of a project which uses Afterglow as a dependency, as
described above, see
[afterglow-max](https://github.com/Deep-Symmetry/afterglow-max#afterglow-max),
which hosts Afterglow inside [Cycling ‘74’s Max](https://cycling74.com/).

## Status

Although Afterglow is far from finished, it&rsquo;s ready for the world to
start exploring, and helping decide directions in which to grow next
(as well as identifying areas where the documentation needs
clarification or reinforcement).

Most of the crazy ideas have panned out and been implemented, and I am
fleshing out the basic details needed for everyday use. The examples
are starting to be intriguing and informative, and the [Developer
Guide](https://afterglow-guide.deepsymmetry.org) is getting
substantial. The modeling of fixtures, channels, etc. is coming
together nicely, though there may be a few more changes.

There is now an embedded web application, which is growing into a show
control interface for people who are not Clojure hackers, and a useful
adjunct to the Ableton Push and Launchpad family control surface
interfaces. Each is explained in the documentation link above.
Afterglow also includes the beginnings of a show visualizer for
designing and working on effects without having to physically hook up
lights (a proof of concept, really, at this point). This is
implemented in WebGL using a volumetric ray tracer and looks quite
promising, at least for a small number of fixtures; it will probably
overwhelm the graphics processor on most systems once you add too many
lights. However, the framework can be used by someone who actually
knows OpenGL programming to build a more scalable preview (albeit one
that probably doesn’t look quite so photo-realistic with beams
impacting drifting fog). This is an area where I would love some help
if it sounds interesting!

## Usage

> The rest of this document primarily provides an introduction to the
> configuration of Afterglow from the command line and text files. The
> show control interface is explained in the
> [web](https://afterglow-guide.deepsymmetry.org/afterglow/README.html#web-ui)
> and
> [Push](https://afterglow-guide.deepsymmetry.org/afterglow/push2.html)
> sections.

Although you will often want to use Afterglow from a Clojure repl, you
can also bring it up as an executable jar, and run it using `java
-jar` with command-line arguments:

```
> java -jar afterglow.jar --help

afterglow 0.2.4, a live-coding environment for light shows.
Usage: afterglow [options] [init-file ...]
  Any init-files specified as arguments will be loaded at startup,
  in the order they are given, before creating any embedded servers.

Options:
  -w, --web-port PORT     16000               Port number for web UI
  -n, --no-browser                            Don't launch web browser
  -o, --osc-port PORT     16001               Port number for OSC server
  -r, --repl-port PORT                        Port number for REPL, if desired
  -l, --log-file PATH     logs/afterglow.log  File into which log is written
  -H, --olad-host HOST    localhost           Host name or address of OLA daemon
  -P, --olad-port PORT    9010                Port number OLA daemon listens on
  -q, --convert-qxf PATH                      Convert QLC+ fixture file and exit
  -h, --help                                  Display help information and exit

If you translate a QLC+ fixture definition file, Afterglow will try to write
its version in the same directory, but won't overwrite an existing file.

If you do not explicitly specify a log file, and Afterglow cannot write to
the default log file path, logging will be silently suppressed.

Please see https://github.com/Deep-Symmetry/afterglow for more information.
```

As noted, you can pass a list of init-files when you run Afterglow
this way, which gives you the opportunity to set up the actual
universes, fixtures, effects, and cues that you want to use in your
show. As a starting point, you could put something like the following
in a file `my-show.clj` and then invoke Afterglow as `java -jar afterglow.jar my-show.clj`:

```clojure
(ns my-show
  "Set up the fixtures, effects, and cues I actually want to use."
  ;; TODO: Your list of required namespaces will differ from this, depending on
  ;;       what fixtures you actually use, and what effects and cues you create.
  (:require [afterglow.core :as core]
            [afterglow.transform :as tf]
            [afterglow.effects.color :refer [color-effect]]
            [afterglow.effects.cues :as cues]
            [afterglow.effects.dimmer :refer [dimmer-effect]]
            [afterglow.effects.fun :as fun]
            [afterglow.effects.movement :as move]
            [afterglow.effects.oscillators :as oscillators]
            [afterglow.effects.params :as params]
            [afterglow.fixtures.blizzard :as blizzard]
            [afterglow.rhythm :as rhythm]
            [afterglow.show :as show]
            [afterglow.show-context :refer :all]
            [com.evocomputing.colors :refer [create-color hue adjust-hue]]
            [taoensso.timbre :as timbre]))

(defonce ^{:doc "Holds my show if it has been created,
  so it can be unregistered if it is being re-created."}
  my-show
  (atom nil))

(defn use-my-show
  "Set up the show on the OLA universes it actually needs."
  []

  ;; Create, or re-create the show. Make it the default show so we don't
  ;; need to wrap everything below in a (with-show sample-show ...) binding.
  (set-default-show!
   (swap! my-show (fn [s]
                    (when s
                      (show/unregister-show s)
                      (with-show s (show/stop!)))
                    ;; TODO: Edit this to list the actual OLA universe(s) that
                    ;;       your show needs to use if they are different than
                    ;;       just universe 1, as below, and change the description
                    ;;       to something descriptive and in your own style:
                    (show/show :universes [1] :description "My Show"))))

  ;; TODO: Replace this to patch in an actual fixture in your show, at its actual
  ;;       universe, DMX address, physical location and orientation, then add all
  ;;       your other fixtures one by one.
  (show/patch-fixture! :torrent-1 (blizzard/torrent-f3) 1 1
                       :x (tf/inches 44) :y (tf/inches 51.75) :z (tf/inches -4.75)
                       :y-rotation (tf/degrees 0))

  ;; Return the show's symbol, rather than the actual map, which gets huge with
  ;; all the expanded, patched fixtures in it.
  '*show*)

(core/init-logging)  ; Log at :info level to rotating files in logs/ subdirectory.
(use-my-show)  ; Set up my show as the default show, using the function above.

;; TODO: Add your custom effects, then assign them to cues with sensible colors
;;       See afterglow.examples for examples.
```

As noted, you will want to look at the
[afterglow.examples](src/afterglow/examples.clj)
namespace for some examples of how to populate this file; the rest of
this section gives an overview and walk-through of how pieces of that
namespace work. The `:require` section at the top of `my-show.clj` is
set up to make it easy to cut and paste from these examples, although
it is not complete, and you will eventually need to learn how to
adjust and optimize it yourself.

> The example code above configures Afterglow to log to a set of
> rotating log files in a `logs/` subdirectory of your project.
> Afterglow will attempt to create that directory if it does not
> exist. If you want to see any logging information, which can be
> quite useful when troubleshooting, you will need to ensure that the
> path to the logs directory is writeable (or that the logs directory
> exists and is writable), otherwise the logging mechanism will
> silently do nothing. The logs will stay out of your way until you
> are interested in them, and take up a limited amount of space, but
> whenever you do want to watch what Afterglow is doing, you can look
> at them, or `tail -f logs/afterglow.log` to watch it live.

As your show gets more complex, you may want to split this into
multiple files, which you can either load by listing them all on the
command line, or by using Clojure&rsquo;s `load-file` function from within
the first file. Or, once you are comfortable with idomatic Clojure
development, by organizing them into a hierarchy of namespaces, and
using the normal `:require` mechanism that is used to pull in
Afterglow&rsquo;s own namespaces.

> :heavy_exclamation_mark: At this early stage of development, using
> Afterglow as an executable jar is less-tested territory, and you may
> find surprising bugs... though this is becoming less of an issue
> since the advent of
> [afterglow-max](https://github.com/Deep-Symmetry/afterglow-max#afterglow-max),
> which is putting Afterglow through its paces as an embedded jar. In
> any case, although the project will gradually evolve into a system
> that non-Clojure hackers can use, for now you are probably best off
> playing with it inside a Clojure development environment, or within
> Max, likely with a Clojure environment connected via nREPL.

Assuming you are using it from within a REPL, there is a namespace
`afterglow.examples` which is intended to help you get started quickly
in exploring the environment, as well as serving as an example of how
to configure your own shows, fixtures, effects, and cues.

> The next two lines are not needed if you are using a checkout of the
> Afterglow source code rather than the library version described
> above, since the project is configured to start you in this
> namespace for convenience.

```clojure
(require 'afterglow.examples)
(in-ns 'afterglow.examples)
```

When you run Afterglow as an executable jar, it will automatically
open a web browser window on its embedded web interface. If you are
using it in another way, you can bring up the web interface, and open
a browser window on it, with a one-liner like this (the first argument
specifies the port on which to run the web interface, and the second
controls whether a browser window should be automatically opened):

```clojure
(core/start-web-server 16000 true)
```

<img alt="Web Interface"
src="doc/modules/ROOT/assets/images/WebHome.png" width="537"
height="427">


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This change log follows the conventions of
[keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]

### Fixed

- Afterglow would crash when told to convert a QLC+ fixture definition
  in the current working directory (with no path component),
  [#72](https://github.com/Deep-Symmetry/afterglow/issues/72).
- External links were not being checked in the developer guide, but
  even so the link checker plugin ended up failing. Switched to
  htmltest, which is used on other Deep Symmetry projects, enabled
  external link checking (although not yet for GitHub source links
  because they are hitting rate limits, there are so many of them),
  and fixed a great many broken/stale links.

### Added

- A new effect builder, `wrap-fade-in-out`, to make it easy to build
  effects that fade in when you start them, and out when you tell them
  to end.
- When resolving dynamic parameters, resolution will be repeated if
  the result was still a dynamic parameter. This allows, for example,
  oscillated parameters to be stored in show variable parameters and
  swapped out during the lifetime of a cue. Resolving the variable
  parameter will first obtain the oscillated parameter, which will be
  resolved again for the appropriate number based on the show
  metronome snapshot.
- The `variable-effect` function now takes optional keyword arguments
  which allow a more meaningful effect name to be specified, and can
  suppress the resolution of dynamic parameters, allowing the raw
  parameter itself to be set into a show variable while the effect is
  active, so that it can be used by other effects.


## [0.2.5] - 2022-01-12

### Fixed

- The built-in converter for QLC+ fixture definitions (`.qxf` files)
  now supports the new XML schema used by QLC+.
- There were some off-by-one errors in the metronome, which canceled
  each other out in the places they were used, but would have caused
  problems in new situations. They were discovered while porting the
  logic to Java for the
  [electro](https://github.com/Deep-Symmetry/electro) library.
- Abrupt tempo changes could cause the metronome to jump to the wrong
  place, even the wrong beat. This is now prevented, leaving you at
  the exact same place in the beat grid before and after any tempo
  change.
- The metronome better handles negative time, and has some improved
  documentation.

### Added

- The detailed documentation has been reorganized as a Developer Guide
  and is now built by [Antora](http://antora.org) to provide easier
  navigation and a more readable presentation.
- The Developer Guide is now available through the built-in web server
  even if you do not have a connection to the internet.

### Changed

- Upgraded to version 2.0 of the Eclipse Publice License.

## [0.2.4] - 2017-04-18

### Fixed

- The pixels of the Blizzard Pixellicious were flipped over the X
  axis.
- The web interface needed to have the Beat Link `DeviceFinder`
  started before trying to render the UI, or it would only partially
  render because of an exception when trying to use it.
- Calling `show/patch-fixture!` returned a huge recursive data
  structure which would cause the REPL to choke trying to print it,
  especially in an editor. Now it just returns the key identifying the
  fixture that was patched.

### Added

- A new command-line argument, `-n`, tells afterglow not to try to
  launch a web browser when it is invoked from the command-line. This
  is to enable headless operation on small servers which lack a
  windowing environment.

## [0.2.3] - 2016-06-09

### Fixed

- Rich controller bindings (Ableton Push and Novation Launchpad
  families) were not being successfully auto-bound under Windows
  because of significant differences in the way their port names were
  assigned on that platform. Fixing that was surprisingly tricky, but
  it has been done. I look forward to someone being able to help test
  this under Linux now.
- The graphical display on the Ableton Push 2 could not be connected
  under Windows due to an issue in the Wayang library. That library
  has been fixed, and the new version is embedded in this release.

## [0.2.2] - 2016-05-30

### Added

- The CoreMIDI4J library is now embedded within Afterglow, so it does
  not need to be separately installed by the user for MIDI to work
  properly on Mac OS X. If you have a separate installation of
  CoreMIDI4J in `/Library/Java/Extensions`, you should remove it to
  avoid version conflicts with newer versions shipped with Afterglow.
- Afterglow now uses the new beat-link library to synchronize with
  Pioneer DJ Link equipment, which enables several new features such
  as tracking the master player, synching to bars as well as beats,
  and having the metronome reflect the current track position.
- Holding down Shift while pressing a scroll arrow on the Launchpad
  family of controllers now moves you as far as possible in that
  direction, as it did on the Push 2.
- Head information is now available to channel effects (which include
  dimmer effects), so they can use dynamic parameters with spatial
  components, just like all the other kinds of effects can.

### Fixed

- When updating large sections of the cue grid colors, we no longer
  send all of them at once, because this was overflowing buffers on
  the Push 2 and losing some updates. Instead we send them in batches,
  ending each batch with a query, and wait for the response so we know
  the controller has caught up.
- Updated to the latest release of the Wayang library for drawing on
  the Push 2 display, which solves an issue that prevented the display
  from working under Window.

### Changed

- The hue and saturation gauges on the Push 2 are now drawn using a
  masking image so they can be anti-aliased to the same outline shape
  as the other gauges, and look much cleaner.
- Boolean gauges on the Push 2 are drawn in red or green for No and

src/afterglow/core.clj
(ns afterglow.core
  "This is the main class for running Afterglow as a self-contained JAR application.
  When you are learning and experimenting in your REPL, the main
  namespace you want to be using is afterglow.examples"
  (:require [afterglow.fixtures.qxf :as qxf]
            [afterglow.init]
            [afterglow.version :as version]
            [afterglow.web.handler :refer [app]]
            [afterglow.web.session :as session]
            [clojure.java.browse :as browse]
            [clojure.tools.cli :refer [parse-opts]]
            [clojure.tools.nrepl.server :as nrepl]
            [environ.core :refer [env]]
            [ola-clojure.ola-client :as ola-client]
            [org.httpkit.server :as http-kit]
            [overtone.osc :as osc]
            [selmer.parser :as parser]
            [taoensso.timbre.appenders.3rd-party.rotor :as rotor]
            [taoensso.timbre :as timbre])
  (:import [java.net InetAddress]
           [org.deepsymmetry.beatlink DeviceFinder])
  (:gen-class))

(defonce ^{:doc "Holds the running web UI server, if there is one, for later shutdown."}
  web-server (atom nil))

(defonce ^{:doc "Holds the running OSC server, if there is one, for later shutdown."}
  osc-server (atom nil))

(defonce ^{:doc "Holds the future which is cleaning up expired web sessions, if any."}
  session-cleaner (atom nil))

(defonce ^{:doc "Holds the running REPL server, if there is one, for later shutdown."}
  nrepl-server (atom nil))

(defn- create-appenders
  "Create a set of appenders which rotate the file at the specified path."
  [path]
  {:rotor (rotor/rotor-appender {:path path
                                 :max-size 100000
                                 :backlog 5})})

(defonce ^{:private true
           :doc "The default log appenders, which rotate between files
           in a logs subdirectory."}
  appenders (atom (create-appenders "logs/afterglow.log")))

(defn- init-logging-internal
  "Performs the actual initialization of the logging environment,
  protected by the delay below to insure it happens only once."
  []
  (timbre/set-config!
   ;; See http://ptaoussanis.github.io/timbre/taoensso.timbre.html#var-*config* for more details.
   {:min-level :info ; #{:trace :debug :info :warn :error :fatal :report}
    :enabled?  true

    :middleware [] ; (fns [data]) -> ?data, applied left->right

    :timestamp-opts {:pattern  "yyyy-MMM-dd HH:mm:ss"
                     :locale   :jvm-default
                     :timezone (java.util.TimeZone/getDefault)}

    :output-fn timbre/default-output-fn ; (fn [data]) -> string
    })

  ;; Install the desired log appenders
  (timbre/merge-config!
   {:appenders @appenders})

  ;; Disable Selmer's template cache in development mode
  (when (Boolean/valueOf (env :dev)) (parser/cache-off!)))

(defonce ^{:private true
           :doc "Used to ensure log initialization takes place exactly once."}
  initialized (delay (init-logging-internal)))

(defn init-logging
  "Set up the logging environment for Afterglow. Called by main when invoked
  as a jar, and by the examples namespace when brought up in a REPL for exploration,
  and by extensions such as afterglow-max which host Afterglow in Cycling '74's Max."
  ([] ;; Resolve the delay, causing initialization to happen if it has not yet.
   @initialized)
  ([appenders-map] ;; Override the default appenders, then initialize as above.
   (reset! appenders appenders-map)
   (init-logging)))

(defn- valid-host?
  "Check whether a string represents a valid host name."
  [name]
  (try
    (InetAddress/getByName name)
    true
    (catch Exception e
      false)))

(defn- println-err
  "Prints objects to stderr followed by a newline."
  [& more]
  (binding [*out* *err*]
    (apply println more)))

(def ^:private log-file-error
  "Holds the validation failure message if the log file argument was
  not acceptable."
  (atom nil))

(defn- bad-log-arg
  "Records a validation failure message for the log file argument, so
  a more specific diagnosis can be given to the user. Returns false to
  make it easy to invoke from the validation function, to indicate
  that validation failed after recording the reason."
  [& messages]
  (reset! log-file-error (clojure.string/join " " messages))
  false)

(defn- valid-log-file?
  "Check whether a string identifies a file that can be used for logging."
  [path]
  (let [f (clojure.java.io/file path)
        dir (or (.getParentFile f) (.. f (getAbsoluteFile) (getParentFile)))]
    (if (.exists f)
      (cond  ; The file exists, so make sure it is writable and a plain file
        (not (.canWrite f)) (bad-log-arg "Cannot write to log file")
        (.isDirectory f) (bad-log-arg "Requested log file is actually a directory")
        ;; Requested existing file looks fine, make sure we can roll over
        :else (or (.canWrite dir)
                  (bad-log-arg "Cannot create rollover log files in directory" (.getPath dir))))
      ;; The requested file does not exist, make sure we can create it
      (if (.exists dir)
        (and (or (.isDirectory dir)
                 (bad-log-arg "Log directory is not a directory:" (.getPath dir)))
             (or (.canWrite dir) ; The parent directory exists, make sure we can write to it
                 (bad-log-arg "Cannot create log file in directory" (.getPath dir))))
        (or (.mkdirs dir) ; The parent directory doesn't exist, make sure we can create it
          (bad-log-arg "Cannot create log directory" (.getPath dir)))))))

(def cli-options
  "The command-line options supported by Afterglow."
  [["-w" "--web-port PORT" "Port number for web UI"
    :default (or (when-let [default (env :web-port)] (Integer/parseInt default)) 16000)
    :parse-fn #(Integer/parseInt %)
    :validate [#(< 0 % 0x10000) "Must be a number between 0 and 65536"]]
   ["-n" "--no-browser" "Don't launch web browser"]
   ["-o" "--osc-port PORT" "Port number for OSC server"
    :default (or (when-let [default (env :osc-port)] (Integer/parseInt default)) 16001)
    :parse-fn #(Integer/parseInt %)
    :validate [#(< 0 % 0x10000) "Must be a number between 0 and 65536"]]
   ["-r" "--repl-port PORT" "Port number for REPL, if desired"
    :default (env :repl-port)
    :parse-fn #(Integer/parseInt %)
    :validate [#(< 0 % 0x10000) "Must be a number between 0 and 65536"]]
   ["-l" "--log-file PATH" "File into which log is written"
    :default (or (env :log-file) "logs/afterglow.log")
    :validate [valid-log-file? @log-file-error]]
   ["-H" "--olad-host HOST" "Host name or address of OLA daemon"
    :default (or (env :olad-host) "localhost")
    :validate [valid-host? "Must be a valid host name"]]
   ["-P" "--olad-port PORT" "Port number OLA daemon listens on"
    :default (or (when-let [default (env :olad-port)] (Integer/parseInt default)) 9010)
    :parse-fn #(Integer/parseInt %)
    :validate [#(< 0 % 0x10000) "Must be a number between 0 and 65536"]]
   ["-q" "--convert-qxf PATH" "Convert QLC+ fixture file and exit"]
   ["-h" "--help" "Display help information and exit"]])

(defn usage
  "Print message explaining command-line invocation options."
  [options-summary]
  (clojure.string/join
   \newline
   [(str (version/title) " " (version/tag) ", a live-coding environment for light shows.")
    (str "Usage: " (version/title) " [options] [init-file ...]")
    "  Any init-files specified as arguments will be loaded at startup,"
    "  in the order they are given, before creating any embedded servers."
    ""
    "Options:"
    options-summary
    ""
    "If you translate a QLC+ fixture definition file, Afterglow will try to write"
    "its version in the same directory, but won't overwrite an existing file."
    ""
    "If you do not explicitly specify a log file, and Afterglow cannot write to"
    "the default log file path, logging will be silently suppressed."
    ""
    "Please see https://github.com/Deep-Symmetry/afterglow for more information."]))

(defn error-msg
  "Format an error message related to command-line invocation."
  [errors]
  (str "The following errors occurred while parsing your command:\n\n"
       (clojure.string/join \newline errors)))

(defn exit
  "Terminate execution with a message to the command-line user."
  [status msg]
  (if (zero? status)
    (println msg)
    (println-err msg))
  (System/exit status))

(defn start-web-server
  "Start the embedded web UI server on the specified port. If a truthy
  value is supplied for browser, opens a web browser window on the
  newly launched server. If the server was already running, logs a
  warning. Either way, makes sure the background thread which cleans
  up expired sessions is running."
  ([port]
   (start-web-server port false))
  ([port browser]
   (swap! web-server #(if %
                        (timbre/warn "Not starting web server because it is already running.")
                        (do
                          (.start (DeviceFinder/getInstance)) ; The web UI wants to know about DJ Link devices
                          (http-kit/run-server app {:port port}))))

   ;;Start the expired session cleanup job if needed
   (swap! session-cleaner #(or % (session/start-cleanup-job!)))

   ;; Launch the browser if requested
   (when browser (browse/browse-url (str "http://localhost:" port)))))

(defn start-osc-server
  "Start the embedded OSC server on the specified port."
  [port]
  (when @osc-server (throw (IllegalStateException. "OSC server is already running.")))
  (reset! osc-server (osc/osc-server port "Afterglow")))

(defn stop-osc-server
  "Shut down the embedded OSC server if it is running."
  []
  #_(osc/zero-conf-off)
  (try
    (swap! osc-server (fn [server]
                        (when server
                          (osc/osc-rm-all-listeners server)
                          (osc/osc-rm-all-handlers server)
                          (osc/osc-close server)
                          nil)))
    (catch Throwable t
      (timbre/error t "failed to shut down OSC server"))))

(defn start-nrepl
  "Start a network REPL for debugging or remote control."
  [port]
  (try
    (swap! nrepl-server #(do (when % (nrepl/stop-server %)) (nrepl/start-server :port port)))
    (timbre/info "nREPL server started on port" port)
    (catch Throwable t
      (timbre/error t "failed to start nREPL"))))

(defn stop-servers
  "Shut down the embedded web UI, OSC and NREPL servers."
  []
  (timbre/info "shutting down embedded servers...")
  (swap! web-server #(do (when % (% :timeout 100)) nil))
  (stop-osc-server)
  (swap! nrepl-server #(do (when % (nrepl/stop-server %)) nil))
  (swap! session-cleaner #(do (when % (future-cancel %)) nil))
  (timbre/info "shutdown complete!"))

(defn -main
  "The entry point when invoked as a jar from the command line. Parse options
  and start servers on the appropriate ports."
  [& args]
  (let [{:keys [options arguments errors summary]} (parse-opts args cli-options)]

    ;; Handle help, error conditions, and fixture definition translation
    (cond
      (:help options) (exit 0 (usage summary))
      errors (exit 1 (str (error-msg errors) "\n\n" (usage summary)))
      (:convert-qxf options) (let [[status message] (qxf/convert-qxf (:convert-qxf options))]
                               (exit status message)))

    ;; Set up the logging environment
    (reset! appenders (create-appenders (:log-file options)))
    (init-logging)

    ;; Load any requested initialization files
    (doseq [f arguments]
      (try
        (timbre/info "Loading init-file" f)
        (binding [*ns* (the-ns 'afterglow.init)]
          (load-file f))
        (catch Throwable t
          (timbre/error t "Problem loading init-file" f)
          (println-err "Failed to load init-file" f)
          (println-err (.getMessage t))
          (println-err "See" (:log-file options) "for stack trace.")
          (System/exit 1))))

    ;; Set up embedded servers
    (.addShutdownHook (Runtime/getRuntime) (Thread. stop-servers))
    (reset! ola-client/olad-host (:olad-host options))
    (when-not (#{"localhost" "127.0.0.1"} @ola-client/olad-host) (ola-client/use-buffered-channel))
    (reset! ola-client/olad-port (:olad-port options))
    (timbre/info "Will find OLA daemon on host" @ola-client/olad-host ", port" @ola-client/olad-port)
    (start-web-server (:web-port options) (not (:no-browser options)))
    (timbre/info "Web UI server on port:" (:web-port options))
    (start-osc-server (:osc-port options))
    (timbre/info "OSC server on port:" (:osc-port options))
    (when-let [port (:repl-port options)]
      (start-nrepl port)
      (timbre/info "nrepl server on port:" port))

    (timbre/info "Startup complete:" (version/title) (version/tag))))
